sound/soc/sof/ipc4-topology.c

```diff
@@ -1430,6 +1499,20 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 		msg = &mixer->msg;
 		break;
 	}
+	case snd_soc_dapm_effect:
+	{
+		struct sof_ipc4_process *process = swidget->private;
+
+		/* ipc config data size should never be 0? */
+		if (!process->ipc_config_size)
+			return -EINVAL;
+
+		ipc_size = process->ipc_config_size;
+		ipc_data = process->ipc_config_data;
```

Comment:
these are never set?

Reply:
@libinyang this PR seems incomplete. What we want is foundation code for initializing any effect-type widget. Take IPC3 for example: A widget's IPC payload consists of 2 parts: 1. The sof_ipc_process structure with common parameters that apply to all effect widgets 2. widget-specific data that comes from topology and is parsed from the kcontrol data.

We should do something similar here for IPC4. Split the IPC payload for effect widgets into 2 parts: 1. common data for all effect widgets 2. widget-specific data (this could include the output format as well if it is optional).

The widget-specific data should ideally come directly in binary format from topology so that the kernel doesnt need to do any special handling and simply append the data to the common data and increase the payload size accordingly.

Also, you might want to check if the output format is strictly optional or can be unconditionally included for all effect widgets. Maybe the FW module code will ignore it if not needed.

Reply:
> @libinyang this PR seems incomplete. What we want is foundation code for initializing any effect-type widget. Take IPC3 for example: A widget's IPC payload consists of 2 parts: 1. The sof_ipc_process structure with common parameters that apply to all effect widgets 2. widget-specific data that comes from topology and is parsed from the kcontrol data.
> 
> We should do something similar here for IPC4. Split the IPC payload for effect widgets into 2 parts: 1. common data for all effect widgets 2. widget-specific data (this could include the output format as well if it is optional).
> 
> The widget-specific data should ideally come directly in binary format from topology so that the kernel doesnt need to do any special handling and simply append the data to the common data and increase the payload size accordingly.
> 

Widget-specific data does comes from topology. And some data should come from kcontrol data, such as WoV blob. Some data is coming from formal tokens, such as input format and output format, just like copier. The config data here is the input format and output format. So this should come from tokens, which is the same with copier. 
To make the code easier to understand. I would like submit the micsel code together with is patch. So we can get know what the config data is.

> Also, you might want to check if the output format is strictly optional or can be unconditionally included for all effect widgets. Maybe the FW module code will ignore it if not needed.

No, we can't include output format if the module doesn't use it. This is because the payload format is various. Some module may use base format + output fomrat and some modules may use base format + extra data. So if output is included, some modules may parse the payload wrongly.